### ISSUE TITLE
Memory leak with batched inference?

### ISSUE THREAD

**Issue by apoorvkh** [ndif-team/nnsight#396]
I am suspicious of a significant memory leak when I am running batched inference with `nnsight`. I am using `torch==2.6.0` and `nnsight=0.4.5` on Python 3.12 on an A40 GPU. This is just a simple demo, but this behavior is unfortunately blocking my actual experiments at a reasonable scale :(

I'll provide two functionally-equivalent scripts as an MVP.

(1) uses standard model inference and only consumes 5 GB CUDA memory at batch size 100.
(2) uses the `tracer.invoke()` API for batching with nnsight and consumes 41 GB of GPU memory at batch size 100!

Moreover, I observe linear memory growth with batch size in case (1), as expected, but perhaps *exponential* growth in case (2).

(1)

```python
import torch
from nnsight import LanguageModel

model = LanguageModel("meta-llama/Llama-3.2-1B", device_map="auto", dispatch=True)

batch_size = 100
inputs = model.tokenizer("Amazon's former CEO attended Oscars", return_tensors='pt').to(device=model.device)
inputs["input_ids"] = inputs["input_ids"].expand(batch_size, -1)
inputs["attention_mask"] = inputs["attention_mask"].expand(batch_size, -1)

torch.cuda.reset_peak_memory_stats()

with torch.no_grad():
    logits = model(**inputs).logits

print(logits.shape)
print(f"{torch.cuda.max_memory_allocated() / 1e9} GB")
```

![Image](https://github.com/user-attachments/assets/5383c7c8-05b3-4e8f-8ff2-9d78cf47047c)

(2)

```python
import torch
from nnsight import LanguageModel

model = LanguageModel("meta-llama/Llama-3.2-1B", device_map="auto", dispatch=True)

batch_size = 100
prompts = ["Amazon's former CEO attended Oscars"] * batch_size

torch.cuda.reset_peak_memory_stats()

logits = []
with torch.no_grad():
    with model.trace() as tracer:
        for i, p in enumerate(prompts):
            with tracer.invoke(p):
                logits.append(model.lm_head.output[0].save())
logits = torch.stack(logits, dim=0)

print(logits.shape)
print(f"{torch.cuda.max_memory_allocated() / 1e9} GB")
```

![Image](https://github.com/user-attachments/assets/8a99f9bb-0efd-4e08-9a9d-69f3609071cc)

**Comment by AdamBelfki3** [comment]
Hi @apoorvkh - thanks for bringing this to our attention. Let's take a look first at how we can solve this issue for your experiment setup.

-- Use case

I understand that you want to run activation patching over a large batch of prompts. To overcome the high memory usage produced by the `Invoker`, you tried to index intervene directly on the batch passed to the `Tracer`. So, let's see how we can make that work (I got this code from your discord thread):

```py
import torch
from nnsight import LanguageModel
import random

model = LanguageModel("meta-llama/Llama-3.2-1B", device_map="auto", dispatch=True)

batch_size = 100
prompts = ["Amazon's former CEO attended Oscars"] * batch_size
patches = torch.rand(batch_size, 2048).to('cuda')
patch_locations = [random.randint(0, 10) for _ in range(batch_size)]

batch_outputs = []
with torch.no_grad():
    with model.trace(prompts):
        for i in range(len(prompts)):
            model.model.layers[patch_locations[i]].output[0][i, -2, :] = patches[i]
            logits = model.lm_head.output[0, -1, :].save()
            batch_outputs.append(logits)
batch_outputs = torch.stack(batch_outputs)
```

1. The result produced by this code is all identical tensors. This is because the `lm_head.output` of the first prompt is repeatedly saved and append to the list, instead of doing for all prompts. This step should be done outside the loop (since we are not using the `.invoke` syntax) and over all the prompts.

```py
with torch.no_grad():
    with model.trace(prompts):
       for i in range(len(prompts)):
            model.model.layers[patch_locations[i]].output[0][i, -2, :] = patches[i]

        logits = model.lm_head.output[:, -1, :].save()
        batch_outputs.append(logits)
```

2. This still doesn't solve the problem as the results, although not the same, are inconsistent. In summary, the remaining issue is that we are defining interventions out of order with the model's computation graph.

e.g. 

```py
model.model.layers[10].output[0][i, -2, :] = patches[I]
model.model.layers[4].output[0][i, -2, :] = patches[i]
```

The patching at layer 4 is essentially ineffective because it happens after the model reached the computation at layer 10.

To solve this, we should be looping over the layers of the model instead of looping over the prompts, since we can't guarantee from there the order of interventions. This is how the final version of the code should look like:

**[SOLUTION]**

```py
patch_locations_dict = {}
for i, l in enumerate(patch_locations):
    if l not in patch_locations_dict:
        patch_locations_dict[l] = []
    patch_locations_dict[l].append(i)

batch_outputs = []

with torch.no_grad():
    with model.trace(prompts):
        for layer in range(model.config.num_hidden_layers):
            if layer in patch_locations_dict:
                for i in patch_locations_dict[layer]:
                    model.model.layers[layer].output[0][i, -2, :] = patches[i]

        logits = model.lm_head.output[:, -1, :].save()
        batch_outputs.append(logits)

batch_outputs = torch.stack(batch_outputs)
```

-- General Issue

From my investigation, the `Invoker` related implementation doesn't seem to be leaking memory per se, but rather exploding the cached memory. We can observe this by looking at the allocated vs reserved memory stats and if emptying the cache has any effect.

```py
import torch
from nnsight import LanguageModel

model = LanguageModel("meta-llama/Llama-3.2-1B", device_map="auto", dispatch=True)

batch_size = 100
prompts = ["Amazon's former CEO attended Oscars"] * batch_size

torch.cuda.reset_peak_memory_stats()

logits = []
with torch.no_grad():
    with model.trace() as tracer:
        for i, p in enumerate(prompts):
            with tracer.invoke(p):
                logits.append(model.lm_head.output[0].save())
logits = torch.stack(logits, dim=0)

print(logits.shape)

print(f"{torch.cuda.memory_allocated() / 1e9} GB") # => 5.310894592 GB
print(f"{torch.cuda.memory_reserved() / 1e9} GB") # => 41.548775424 GB

torch.cuda.empty_cache()

print(f"{torch.cuda.memory_allocated() / 1e9} GB") # => 5.310894592 GB
print(f"{torch.cuda.memory_reserved() / 1e9} GB") # => 5.322571776 GB
```

So, what causes the `Invoker` to use so much memory in the first place? It's the way batched interventions are handled by the `InterventionProtocol`.

During the interleaved execution of the model, each invoker-specific activation is narrowed so that interventions can happen on it independently Then it gets concatenated with the rest of the prompts' activations (source: https://github.com/ndif-team/nnsight/blob/2a975a18ac2cdee95158333c19d045661b0be0d8/src/nnsight/intervention/protocols/intervention.py#L181C21-L187C22). The `torch.cat` operation is expensive and creates a new tensor, resulting in more memory being allocated; this can become exponential as the number of concatenations/Invokers increases.

**[SOLUTION]**

Instead of concatenating the result from the Invoker value with the rest of the activations, we should try index setting directly into the original tensor.

Thanks for your thorough issue description, hope this is helpful!

**Comment by apoorvkh** [comment]
Thanks so much for your response!! This is extremely helpful. In order:

1. You're right, I did try this, but typed it in wrong on Discord! My bad.

```python
for i in range(len(prompts)):
    model.model.layers[patch_locations[i]].output[0][i, -2, :] = patches[i]
logits = model.lm_head.output[:, -1, :].save()
```

2. I now understand that I need to patch (within the batch) in the order of the layers. I didn't realize that mattered! I think my code is working after adjusting this!
I'm wondering if this would have mattered if I implemented the hooks manually in PyTorch. For the sake of the library, would it be possible to permit such an arbitrary patching order and rearrange the graph of interventions before the forward pass?

3.
> the Invoker is exploding the cached memory

Unfortunately, I don't quite understand how `nnsight` works under-the-hood, so I'm not able to follow your solution :(

But anyway, now I'm using the following `.invoke()`-free batching pattern. The following script consumes 15 GB GPU memory when `LAYERS_TO_PATCH = range(1)`, but 30 GB when `LAYERS_TO_PATCH = range(10)`. In my view, these loops should be independent (besides w.r.t. `logits`).

Is that a reasonable assumption? Is there a more proper way to write this loop? Thanks so much once again!


```python
LAYERS_TO_PATCH = range(1)

import torch
from nnsight import LanguageModel

model = LanguageModel("meta-llama/Llama-3.2-3B", device_map="auto", dispatch=True)

batch_size = 100
prompt = " ".join(["Amazon's former CEO attended Oscars"] * 5)  # 31 tokens

torch.cuda.reset_peak_memory_stats()
logits = {}

with torch.no_grad():
    for l in LAYERS_TO_PATCH:
        with model.trace([prompt] * batch_size) as tracer:
            model.model.layers[l].output[0][:, -1, :] = torch.rand(batch_size, model.lm_head.in_features, device=model.device)
            logits[l] = model.lm_head.output.save()
        torch.cuda.empty_cache()

print(f"{torch.cuda.max_memory_allocated() / 1e9} GB")
```
